### DISCUSSION TITLE
Resolve Merge Conflicts

### DISCUSSION THREAD

**Discussion by SyntaxisTaxsyn** [gitextensions/gitextensions#8227] [General]
Hi. Great project by the way. For what i wanted to do it has almost ticked all the boxes i needed. 1 Small problem i seem to be having is with the configuration/usage of the custom merge tool. What i did was
1) created a master branch
2) added a file and then committed it
3) made a copy of the branch
4) checked this out and made a change and committed it
5) checked out the master branch, made a change to it and committed it
6) now i tried to merge the branches and it fails like i expect
7) i'm prompted to resolve the merge conflicts manually
8) In the resolve merge conflicts dialog, i click on open my custom merge tool
9) it fails with invalid command line options
10) I double check the setup of the custom merge tool and it is correct (in fact its the same as diff tool but with 1 extra command line switch -m for merge)
11) When i check the command lines passed to the application instead of using a temp file like the diff tool (which is working), instead is passed the name of the file appended with .local or .remote

From this i derive the correct file names are not being passed to the merge tool hence it fails on launch. Either I'm doing it wrong, have configured it wrong, am trying to do something that's not normally possible (could be as I'm fairly new to git), or i have found a bug.

If i open the 2 files with the diff tool option, it works as expected and i can even switch to the merge option within the tool, perform the data merge, save as the current file in the head, and delete the branch, which does the same thing as i wanted. My OCD wants it to work properly though so I would appreciate any help i can get.

I didn't want to post it as an issue yet in case I'm doing something silly so thought i would check it here first. 

Best regards

Edit - I used a tool called process view or so that let me see which command line parameters had been passed to the executable on run, what I saw was clearly the wrong information as a search through the working directory found no such files (that’s what the merge tool was complaining about I think). 
I used the same parameter strings as in the diff tool setup $Local and $Remote, are these paths also valid for merge tool setup?

**Comment by gerhardol** [comment]
Use the Git Command Log (F12) to see the parameters used to the mergetool
If you use the "mergetool" button GE provides the files to be merged to Git and Git deals with temporary files etc, which seem to be what you are seeing here, similar to how difftool is invoked

With the topmost "open with xxx" GE replaces the commandline arguments internally, but the call to the mergetool itself should be the same

**Reply by SyntaxisTaxsyn** [reply]
Ok so a bit more digging later and the problem is a little clearer. It seems that depending on where you select the compare option from within the GUI, it launches the external diff tool using different arguments to suit the location of the files i.e sometimes it puts them in appdata/local/temp sometimes appdata/local/temp\left and depending it comes with or without the name of the actual file on the end of the path, depending on the way the diff tool was launched. 

Now i started writing a wrapper tool to pick up on this and parse the command lines, generate the correct paths for the files and pass them to the diff tool on invoke. I seem to be on the right path and it was working until i tried to diff on an unstaged commit. The second path in the args only supplies the name of the file in the working directory with no context as to where it is for my external diff tool to start. 

The question is, is there a global variable that i can use in the diff tool settings to pass this location on to the diff wrapper tool? From an example i found on the internet i can use $local and $remote to get the paths of the files to compare in the diff tool setup. Is there any other variables i can use in here which specify the path to the working directory so i can resolve this in the wrapper tool?

I tried looking in the user docs but a search on $Local/$Remote turned up nothing.

For reference the setup of my config for difftool args looks like this 

"C:/MergeToolWrapper/MergeToolWrapper.exe" "$LOCAL" "$REMOTE" "D"

Thanks in advance

**Reply by RussKie** [reply]
> Is there any other variables i can use in here which specify the path to the working directory so i can resolve this in the wrapper tool?

You'd need to look at [git help](https://git-scm.com/docs/git-difftool#Documentation/git-difftool.txt---toollttoolgt) and giff tool of your choice docs. 

AFAIK (and I haven't looked at that for awhile, so may be misremembering things), there are generally only those variables that we have. Some tools may accept something else, but this would be an exception rather than a norm.
